csrc/core/scalar_type.hpp

```diff
@@ -79,19 +79,74 @@ class ScalarType {
   NanRepr const nan_repr;         // how NaNs are represented
                                   // (not applicable for integer types)
 
-  int64_t size_bits() const { return mantissa + exponent + is_signed(); }
-  bool is_signed() const { return signed_; }
-  bool is_integer() const { return exponent == 0; }
-  bool is_floating_point() const { return exponent > 0; }
-  bool is_ieee_754() const {
+  // unique id for this scalar type that can be computed at compile time for
+  //  c++17 tempalate specialization this is not needed once we migrate to c++20
+  //  and can pass literal classes as template parameters
+  constexpr int64_t id() const {
```

Comment:
Maybe worth making an alias for `using ScalarTypeId = int64_t;`

Reply:
Good, idea done!